gadgets/deadlock/program.bpf.c

```diff
@@ -8,55 +8,211 @@
+		edge_value.tid = tid;
+		edge_value.mntns_id = mtns_id;
+		edge_value.mutex1_stack_id = held_mutexes[i].stack_id;
+		edge_value.mutex2_stack_id = stack_id;
+		bpf_get_current_comm(&edge_value.comm, sizeof(edge_value.comm));
+
+		int result = bpf_map_update_elem(&edges, &edge_key, &edge_value,
+						 BPF_ANY); // update graph
+		if (result) {
+			bpf_printk(
+				"could not add edge key with mutexes %p and %p. error: %d\n",
+				edge_key.mutex1, edge_key.mutex2, result);
+			continue; // out of memory
+		}
+	}
+	if (!added_mutex) {
+		bpf_printk("could not add mutex %p\n", mutex);
+		return 1; // no more free space on `held_mutexes`
+	}
+	return 0;
+}
+
+/*
+ * Removes mutex from the list of held mutexes for the current thread.
+ *
+ * We don't remove the edges associated with the mutex from the mutex wait graph
+ * as we are detecting "potential" deadlocks (even if they might never happen).
+ *
+ * If we remove the edges, we only detect deadlocks that actually happen during the trace.
+ * But a deadlock that didn't happen during the trace could happen in the future
+ * as it depends on factors such as thread scheduling and order of mutex acquisitions/releases.
+ */
+static __always_inline int trace_mutex_release(struct pt_regs *ctx,
+					       u64 mutex_addr)
+{
+	u64 mtns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mtns_id))
 		return 0;
 
-	gadget_process_populate(&event->proc);
-	event->mutex_addr = mutex_addr;
-	event->operation_raw = operation;
-	event->timestamp_raw = bpf_ktime_get_ns();
+	u32 tid = (u32)bpf_get_current_pid_tgid();
+	u64 mutex = mutex_addr;
 
-	gadget_submit_buf(ctx, &events, event, sizeof(*event));
+	// Fetch the held mutexes for the current thread
+	struct held_mutex *held_mutexes =
+		bpf_map_lookup_elem(&thread_to_held_mutexes, &tid);
+	if (!held_mutexes) {
+		/*
+         * If `held_mutexes` doesn't exist for the tid, then it means we either missed
+         * the acquire event, or were out of memory when adding it.
+         */
+		bpf_printk(
+			"could not find the thread's held mutexes. thread: %d, mutex: %p\n",
+			tid, mutex);
+		return 1;
+	}
 
+	int is_cleared =
+		1; // flag indicating whether all held mutexes by the current thread are cleared
+	for (int i = 0; i < MAX_HELD_MUTEXES; ++i) {
+		// Find the current mutex and clear it
+		if (held_mutexes[i].mutex == mutex) {
+			held_mutexes[i].mutex = 0;
+			held_mutexes[i].stack_id = 0;
+		}
+		if (held_mutexes[i].mutex != 0 && held_mutexes[i].mutex != 0) {
+			// Thread still holds a mutex
+			is_cleared = 0;
+		}
+	}
+	if (is_cleared) {
+		// Remove the entry from the map if no mutex is held by the thread
+		bpf_map_delete_elem(&thread_to_held_mutexes, &tid);
+	}
 	return 0;
 }
 
 /* mutex acquisition */
 SEC("uprobe/libc:pthread_mutex_lock")
 int BPF_UPROBE(trace_uprobe_mutex_lock, void *mutex_addr)
 {
-	return gen_mutex_event(ctx, lock, (u64)mutex_addr);
+	return trace_mutex_acquire(ctx, (u64)mutex_addr);
```

Comment:
You could add a filter by pid:

```c
const volatile pid_t targ_pid = 0;
GADGET_PARAM(targ_pid);
...
if (targ_pid && targ_pid != pid)
    return 0;
```

That would be useful when the process is not in a container but on the host, and the users does not want to check all processes on the host but only one:
```
sudo ig run ghcr.io/inspektor-gadget/gadget/deadlock --host --targ-pid=1234
```

Reply:
I and @burak-ok discussed this in our last call and decided to implement it in a future PR (given the footprint of this PR). Or should I do it in this PR itself?

Reply:
That's fine for me to postpone to another PR.